tests/Test.ADAL.NET.Unit/AdalDotNetTests.cs

```diff
@@ -1494,5 +1496,84 @@ public void Initialize()
             }
         }
 
+        [Description("Telemetry tests")]
+        public void Telemetry()
+        {
+            Microsoft.IdentityModel.Clients.ActiveDirectory.Telemetry telemetry =
+            Microsoft.IdentityModel.Clients.ActiveDirectory.Telemetry.GetInstance();
+            Assert.IsNotNull(telemetry);
+
+            string requestIDOne = telemetry.RegisterNewRequest();
+            Assert.IsNotNull(requestIDOne);
+
+            telemetry.StartEvent(requestIDOne, "acquire_token");
+            string requestIDTwo = telemetry.RegisterNewRequest();
+            Assert.AreEqual(telemetry.EventsStored(), 1);
+
+            DefaultEvent testDefaultEvent = new DefaultEvent("random_event");
+            Assert.IsNotNull(testDefaultEvent.ApplicationName);
+            Assert.IsNotNull(testDefaultEvent.ApplicationVersion);
+            Assert.AreEqual(testDefaultEvent.Tenant,"tid");
+            Assert.AreEqual(testDefaultEvent.PasswordChangeUrl, "pwd_url");
+            Assert.AreEqual(testDefaultEvent.PasswordExpiration, "pwd_exp");
+            Assert.AreEqual(testDefaultEvent.Idp,"idp");
+            Assert.AreEqual(testDefaultEvent.Issuer,"iss");
+
+            DispatcherImplement dispatcher = new DispatcherImplement();
+            telemetry.RegisterDispatcher(dispatcher,true);
+            dispatcher.clear();
+            string requestIDThree = telemetry.RegisterNewRequest();
+            telemetry.StartEvent(requestIDThree, "event_3");
+            telemetry.StopEvent(requestIDThree,testDefaultEvent,"event_3");
+            Assert.AreEqual(dispatcher.Count,1);
+            //dispatcher.file();
+
+            telemetry.StartEvent(requestIDTwo, "cache_lookup");
+            DefaultEvent cacheDefaultEvent = new DefaultEvent("cache_lookup");
+            telemetry.StopEvent(requestIDTwo, cacheDefaultEvent, "cache_lookup");
+            Assert.AreEqual(dispatcher.Count,2);
+
+            string requestIDFour = telemetry.RegisterNewRequest();
+            telemetry.StartEvent(requestIDFour, "crypto_event");
+            DefaultEvent cryptoDefaultEvent = new DefaultEvent("crypto_event");
+            telemetry.StopEvent(requestIDFour, cacheDefaultEvent, "crypto_event");
+            Assert.AreEqual(dispatcher.Count, 3);
+
+            dispatcher.file();
+        }
+
+        class DispatcherImplement : IDispatcher
+        {
+            private readonly List<List<Tuple<string,string>>> storeList = new List<List<Tuple<string, string>>>();
+
+            void IDispatcher.Dispatch(List<Tuple<string, string>> Event)
+            {
+                storeList.Add(Event);
+            }
+
+            public int Count
+            {
+                get { return storeList.Count; }
+            }
+
+            public void clear()
+            {
+                storeList.Clear();
+            }
+
+            public void file()
+            {
+                using (TextWriter tw = new StreamWriter("C:/Users/abgun/test.txt"))
```

Comment:
Removed my specific machine path.

---
In reply to: [74002780](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/494#discussion_r74002780) [](ancestors = 74002780)